### PR TITLE
ci(release): batch release-please commit scans

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- Batch release-please history requests in groups of 25 instead of one commit per request.
- Raise the release-history drift guard to 400 commits while preserving 100 commits of headroom below the 500-commit search limit.
- Keep the release floor pinned to the latest code-scan-action release so unreleased component changes are never skipped.
- Add focused regression coverage for the bounded batch size and safe drift threshold.

## Why

The latest scheduled `release-please-sha-drift` run failed because the code-scan-action 0.1.8 release commit was already 292 commits behind `main`, exceeding the previous 250-commit guard. Current `main` is 299 commits beyond that same safe release floor, and the refreshed PR head is 304 commits beyond it. Advancing the floor would silently discard unreleased code-scan-action changes, so batching reduces roughly 300 individual GraphQL history requests to about 12 requests while the 400-commit guard preserves 100 commits of safety below release-please's 500-commit search limit.

## Verification

- Reproduced both regression failures against current `main`: `commit-batch-size` is `1`, not `25`, and `MAX_DRIFT` is `250`, not `400`.
- `node_modules/.bin/vitest run test/release-please.test.ts test/package-manifests.test.ts --configLoader runner --no-cache` — 17 tests passed.
- `node_modules/.bin/tsc --noEmit --ignoreConfig --target es2022 --module nodenext --moduleResolution nodenext --skipLibCheck --types node test/release-please.test.ts`.
- Executed the actual drift guard against the refreshed head: `304 <= 400`.
- Confirmed `last-release-sha` still equals `code-scan-action-0.1.8` and is an ancestor of the current Promptfoo release.
- `npm run l`, `npm run f`, targeted Biome/Prettier checks, `jq empty release-please-config.json`, and `git diff --check origin/main...HEAD`.

### CODE REVIEW

**Review by jameshiester-oai — COMMENTED** [review]
Reviewed base `cfa9bee146bd38fcabc1c4241ede8889fa95e302` through head `64ce250c459d8b55520cf282f7d3c7c7da8c6cfa`.

Scope: fresh full review of the two-commit PR, including both changed files, the related release-please contract tests and workflow, the pinned upstream release-please implementation, the two cited failed release runs, all existing review threads, and exact-head CI.

Verdict: no new actionable defect survived validation. The batch-size change is wired to upstream GraphQL pagination, the adjusted drift guard accepts the current 255-commit window, and all 45 exact-head checks are green.

Verified findings by severity:
- P0/P1/P2/P3: none newly identified.

Existing unresolved feedback:
- The existing regression-test thread remains applicable and is not duplicated here. `test/release-please.test.ts` accepts every value from 1 through 25, so the known-problematic value 1 still passes. Pinning 25, or enforcing an evidence-backed minimum effective batch size, would protect the behavior this PR changes.
- The earlier drift-guard thread is stale at this head: the follow-up commit raises `MAX_DRIFT` to 300, and the exact-head workflow reports 255 <= 300.

Simplification and test opportunities:
- No material code simplification is needed.
- The most useful additional coverage is the already-raised batch-size regression assertion. Boundary tests for drift 300/301 would also make the shell guard easier to maintain.

Validation performed:
- `source ~/.nvm/nvm.sh && nvm use` -> Node 24.18.0, npm 11.16.0.
- `npm ci --ignore-scripts` -> exact-lock dependency restore completed.
- `npx vitest test/release-please.test.ts --run` -> 1 file, 3 tests passed.
- `npm run tsc` -> passed.
- `jq empty release-please-config.json` -> passed.
- Prettier check for the changed workflow -> passed.
- `git diff --check cfa9bee146bd38fcabc1c4241ede8889fa95e302...64ce250c459d8b55520cf282f7d3c7c7da8c6cfa` -> passed.
- Exact-head `check-drift` log -> 255 commits behind, threshold 300, passed.
- Upstream source inspection confirmed `commit-batch-size` controls GraphQL history page size; independent read-only probes at batch sizes 1, 10, and 25 succeeded and showed fewer aggregate requests at larger batches.

Limitations:
- The requested `/pr-review-toolkit:review-pr all parallel` command was unavailable (exit 127), so I applied its six review lenses manually and reconciled three independent read-only specialist reviews.
- No credentialed end-to-end release-please run was executed. The historical failures occurred on the first GraphQL request, so the exact production root cause is not conclusively established even though direct pagination probes support the change.

**Comment by mldangelo-oai** [comment]
@codex review

File: test/release-please.test.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ import { describe, expect, it } from 'vitest';
 
 const REPO_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
 const MAX_COMMIT_BATCH_SIZE = 25;
+const MAX_RELEASE_HISTORY_SEARCH_COMMITS = 500;
```
Comment (jameshiester-oai) [inline test/release-please.test.ts:12]: **[P2] Pin the runtime search depth instead of duplicating its default**

This `500` is only a test constant: `release-please-config.json` does not set `commit-search-depth`. The pinned action currently bundles release-please 17.6.0, whose implicit default is 500, but a future action SHA upgrade can change that runtime default while this assertion and `MAX_DRIFT=400` still pass. If the real depth drops below the allowed drift, release-please stops before the oldest code-scan release boundary and omits the older unreleased component commits from its split. Set `commit-search-depth: 500` explicitly in the config and derive this headroom assertion from that configured value so the test controls the production contract.

**Review by jameshiester-oai — COMMENTED** [review]
## Review summary

**Reviewed base:** `9ab178ec2772d72bacd533f2dc11ba66208606a5`  
**Reviewed head:** `8222554c92f96e9bd9bce0331ca97ad99750c376`  
**Scope:** Incremental review from authenticated completed ancestor `64ce250c459d8b55520cf282f7d3c7c7da8c6cfa`, plus the complete current base...head PR diff and merged-main interaction.

### Verdict

One new P2 release-integrity test/config gap. The functional change works at this head: `commit-batch-size: 25` feeds GraphQL `history(first: 25)`; the current pinned action v5.0.0 bundles release-please 17.6.0 with a 500-commit default search depth; and the exact workflow reports the oldest release floor 296 commits behind this head, below `MAX_DRIFT=400`.

All 45 current status checks are successful. The exact `check-drift` job passed at 296 <= 400, and Node 24 CI passed the four focused release-please tests plus the 22,005-test suite.

### Verified findings

- **P2:** The new headroom test hard-codes release-please's implicit 500-commit default without pinning `commit-search-depth` in runtime config. A later action SHA upgrade can change the real limit while this test and the 400-commit guard stay green, allowing the code-scan release boundary to fall outside the actual scan and older unreleased component commits to be omitted. See the inline comment.

### Existing feedback

- The resolved batch-size regression thread is fixed: the test now requires exactly 25, so the former value 1 fails.
- The resolved drift-threshold thread is fixed for the current head: the live workflow reports 296 <= 400.
- No existing thread covers the runtime/test coupling issue, so the inline comment is not a duplicate.

### Simplification and test opportunities

The smallest durable design is to own the search depth in `release-please-config.json` and derive the headroom assertion from that configured value. This removes a disconnected test constant and makes future action upgrades unable to silently change the contract.

### Validation

- `/pr-review-toolkit:review-pr all parallel` — unavailable on this host (exit 127); all six lenses were applied manually and reconciled with three bounded read-only specialist passes.
- `git merge-base --is-ancestor 64ce250c459d8b55520cf282f7d3c7c7da8c6cfa HEAD` — passed.
- Exact workflow logic against this head — `af7ec0b…` is 296 commits behind, threshold 400, passed.
- Exact-head `check-drift` job `90122653964` — passed and logged 296 <= 400.
- Exact-head Node 24 job `90122681547` — `test/release-please.test.ts` 4/4 passed; full suite 22,005 passed and 10 skipped.
- Pinned upstream inspection — action v5.0.0 locks release-please 17.6.0; its manifest default is 500 and it passes `commit-batch-size` to GraphQL pagination.
- YAML/JSON contract probes — parsed `MAX_DRIFT: 400` and `commit-batch-size: 25`.
- Prettier check for all three PR-owned files — passed.
- `jq empty release-please-config.json` and `git diff --check` — passed.

### Limitations

Local focused Vitest could not start because this isolated worktree lacked dependencies. The required `npm ci --ignore-scripts` recovery was blocked by the configured Socket policy for newly published `postcss@8.5.23`; exact-head CI supplied the focused and full-suite evidence instead. No credentialed release-please run was executed, and no source, cache, database, or PR state beyond this COMMENT review was modified.

**Comment by mldangelo-oai** [comment]
@codex review

**Review by jameshiester-oai — COMMENTED** [review]
Reviewed base `26b725bd9496351ef269380c9fd83b9c2c338a0e` through head `e2b99235899233ef735ea29fc1ef04b4b7741f3c`. Scope was incremental from authenticated completed ancestor `8222554c92f96e9bd9bce0331ca97ad99750c376`, plus the complete current base...head PR diff and merged-main interaction. The only new commit is a merge of current main, and all three PR-owned blobs are byte-identical to the previously reviewed head.

Verdict: no new actionable defect survived validation. The current release floor is reachable and 304 commits behind this head, so the production drift check passes `304 <= 400`; batch size 25 reduces that window to approximately 13 history requests. All 45 reported exact-head checks are successful.

Verified findings

- P0/P1/P2/P3: no new findings.

Existing unresolved feedback

- The existing **P2** at https://github.com/promptfoo/promptfoo/pull/10160#discussion_r3661830929 remains applicable and is not duplicated. `test/release-please.test.ts` assumes a 500-commit runtime search depth, while `release-please-config.json` still omits `commit-search-depth`. The pinned action currently bundles release-please 17.6.0, whose default is 500, but a future action update can change the runtime default while the 400-commit guard and repository test remain green. Pinning the runtime depth and deriving the test from config is still the smallest durable correction.
- The former batch-size and 250/300 drift-threshold issues remain fixed: configuration and tests now require 25 and 400 respectively.

Simplification and test opportunities

- No additional simplification is warranted beyond resolving the existing search-depth/config coupling. The exact-value batch assertion and named workflow-step assertion cover the changed behavior.

Validation performed

- `npx vitest run test/release-please.test.ts test/package-manifests.test.ts --sequence.seed=10160` — 17/17 passed.
- Three independent specialist passes also ran the focused 4-test release suite and the 17-test combined suite; all passed.
- Exact drift probe — `last-release-sha` exists, is an ancestor, is tagged `code-scan-action-0.1.8`, and is 304 commits behind head; batch size is 25 and `MAX_DRIFT` is 400.
- Pinned upstream verification — action SHA `45996ed1f6d02564a971a2fa1b5860e934307cf7` locks release-please 17.6.0; its manifest source defines the implicit 500 default and reads the optional `commit-search-depth` config.
- `jq empty release-please-config.json`, Prettier check on all three PR files, and `git diff --check` — passed.
- Exact-head CI — 45/45 reported checks successful, including check-drift, platform tests, builds, style, docs, smoke/integration, and CodeQL.
- Worktree remained clean.

Limitations

- No credentialed release-please action run was executed because it can create releases or pull requests. The read-only drift workflow and its current CI run were exercised instead.
- The requested `/pr-review-toolkit:review-pr all parallel` command is unavailable in this environment (exit 127), so the closest read-only six-lens specialist fallback was used and all candidates were independently reconciled.

**Comment by mldangelo-oai** [comment]
@codex review
